haystack/nodes/file_converter/markdown.py

```diff
@@ -40,32 +72,55 @@ def convert(
             attributes. If you want to ensure you don't have duplicate documents in your DocumentStore but texts are
             not unique, you can modify the metadata and pass e.g. `"meta"` to this field (e.g. [`"content"`, `"meta"`]).
             In this case the id will be generated by using the content and the defined metadata.
+        :param remove_code_snippets: Whether to remove snippets from the markdown file.
+        :param extract_headlines: Whether to extract headings from the markdown file.
         """
         if id_hash_keys is None:
             id_hash_keys = self.id_hash_keys
+        if remove_code_snippets is None:
+            remove_code_snippets = self.remove_code_snippets
+        if extract_headlines is None:
+            extract_headlines = self.extract_headlines
+
         with open(file_path, encoding=encoding, errors="ignore") as f:
             markdown_text = f.read()
-        text = self.markdown_to_text(markdown_text)
+
+        # md -> html -> text since BeautifulSoup can extract text cleanly
+        html = markdown(markdown_text)
+
+        # remove code snippets
+        if remove_code_snippets:
+            html = re.sub(r"<pre>(.*?)</pre>", " ", html, flags=re.DOTALL)
+            html = re.sub(r"<code>(.*?)</code>", " ", html, flags=re.DOTALL)
+        soup = BeautifulSoup(html, "html.parser")
+
+        if extract_headlines:
+            text, headlines = self._extract_text_and_headlines_from_soup(soup)
+            if meta is None:
+                meta = {}
+            meta["headlines"] = headlines
+        else:
+            text = soup.get_text()
+
         document = Document(content=text, meta=meta, id_hash_keys=id_hash_keys)
         return [document]
 
-    # Following code snippet is copied from https://gist.github.com/lorey/eb15a7f3338f959a78cc3661fbc255fe
     @staticmethod
-    def markdown_to_text(markdown_string: str) -> str:
+    def _extract_text_and_headlines_from_soup(soup: BeautifulSoup) -> Tuple[str, List[Dict]]:
         """
-        Converts a markdown string to plaintext
-
-        :param markdown_string: String in markdown format
+        Extracts text and headings from a soup object.
         """
-        # md -> html -> text since BeautifulSoup can extract text cleanly
-        html = markdown(markdown_string)
+        headline_tags = {"h1", "h2", "h3", "h4", "h5", "h6"}
```

Comment:
Make the depth level configurable?

Reply:
Or even better - find the deepest tree level in one pass of regular expressions?

Reply:
These are pre-defined HTML tags as listed for example [here](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/Heading_Elements?retiredLocale=de). 
Before converting a markdown file to text, we convert it to HTML. The `headline_tags` set here is just used to check if an HTML element is a headline.